source/extensions/filters/http/composite/action.h

```diff
@@ -39,8 +40,27 @@ class ExecuteFilterActionFactory
             composite_action.typed_config());
     ProtobufTypes::MessagePtr message = Config::Utility::translateAnyToFactoryConfig(
         composite_action.typed_config().typed_config(), validation_visitor, factory);
-    auto callback = factory.createFilterFactoryFromProto(*message, context.stat_prefix_,
-                                                         context.factory_context_);
+
+    Envoy::Http::FilterFactoryCb callback = nullptr;
+    try {
+      if (context.server_factory_context_.has_value()) {
+        callback = factory.createFilterServerFactoryFromProto(
+            *message, context.stat_prefix_, context.server_factory_context_.value());
+      }
+    } catch (EnvoyException& e) {
+      // Instead of causing the crash, the exception is handled gracefully: log the error and
```

Comment:
Why would it cause the crash? Isn't this code executed during configuration parsing? If so it should just reject the config, which is the right action here.

Reply:
Here is actually not the invalid config case. Here is in composite filter 1) first we try to create the delegated filter callbackfrom server factory context 2) if this filter doesn't support this mechanism, we catch the envoyException that has been thrown and fallback to create the filter from factory context.  Existing `CompositeFilterIntegrationTest` that uses set-reponse-code filter is a good example of this. And we don't want to fail here.

I updated the comment to clarify this behavior. PTAL. Thanks for the review!

Reply:
I see. I think the logic is not quite right though. If the FilterFactoryContext is available we should just create a factory using it and not even try using the ServerFactoryContext. All filters must support FilterFactoryContext so it should always work.

If the FilterFactoryContext is not available when the factory should be attempted to be created using the ServerFactoryContext and exception is thrown to reject the config in this case.

Reply:
I think the logic is right but with a slight difference. 
Here is summary: (`create from FactoryContext` is always supported)
1. `ServerFactoryContext` is available +  `FactoryContext` is NOT available + `create from ServerFactoryContext` is supported: 
   -  Create from `ServerFactoryContext`. (The functionality this PR introduce)
2. `ServerFactoryContext` is available + `FactoryContext` is available + `create from ServerFactoryContext` is NOT supported
   - Create from FilterFactoryContext. (Existing behavior) 
3. `ServerFactoryContext` is available + `FactoryContext` is NOT available + `create from ServerFactoryContext` is NOT supported.
    - Release Assert triggered. 
4.  `ServerFactoryContext` is available + `FilterFactoryContext` is available + `Create from ServerFactoryContext` is supported
    - Create from `serverFactoryContext`  since both are supported, either should just works
5. `ServerFactoryContext` is not available: 
   - Create from `FilterFactoryContext` (Existing behavior)

So the difference is in case (4), where `serverFactoryContext`  or `factoryContext` take precedence

Reply:
Even though there might be no functional difference,  I agree with you that creating from `factoryContext` should take precedence as it is the method every filter support. I will change the code the something like below (which is a bit more concise)
```
if (context.factory_context_.has_value()) {
      callback = factory.createFilterFactoryFromProto(*message, context.stat_prefix_,
                                                      context.factory_context_.value());
    } else if (context.server_factory_context_.has_value()) {
      callback = factory.createFilterServerFactoryFromProto(
          *message, context.stat_prefix_, context.server_factory_context_.value());
    } else {
      throw EnvoyException("one of `factory_context_` and `server_factory_context_` must be set in "
                           "HttpFilterActionContext");
    }
 ```
 Thanks for the review!

Reply:
I have added and tested new test filter which mock the ext_proc behavior, but I realized that if I make the changes like above, the method I added in this PR(creating from `serverFactoryContext`) cannot be tested.

Please note: whether `serverFactoryContext` or `factoryContext` take precedence only affects composite filter case.

Reply:
The test filter does not need to have any behavior other than having ability to be created using server context.

Reply:
yes, it does have that ability. But the problem is that ability can not be hit unless 1) let server method take precedence and fallback like here OR 2) still factory method take precedence but change composite filter code(where setup the context) to have empty factory and non-empty server factory. 2) is not good as that basically limit composite filter to filters which support server method(only 1 here).

We can hit this path as normal after that per route PR(where only server context is available) is in. But in the same time that pr requires this PR to create filter 

To solve this dilemma，i am wondering 1) can we merge the code as it is. It will be non functional breakage because there is fallback to factory method 2) Add TODO and change it factory-precedence style code I mentioned above later?